Sources/TuistSigning/GraphMappers/SigningMapper.swift

```diff
@@ -0,0 +1,122 @@
+import Foundation
+import TSCBasic
+import TuistCore
+import TuistSupport
+
+enum SigningMapperError: FatalError, Equatable {
+    case appIdMismatch(String, String, String)
+
+    var type: ErrorType {
+        switch self {
+        case .appIdMismatch:
+            return .abort
+        }
+    }
+
+    var description: String {
+        switch self {
+        case let .appIdMismatch(appId, developmentTeamId, bundleId):
+            return "App id \(appId) does not correspond to \(developmentTeamId).\(bundleId). Make sure the provisioning profile has been added to the right target."
+        }
+    }
+}
+
+public class SigningMapper: GraphMapping {
+    private let signingFilesLocator: SigningFilesLocating
+    private let rootDirectoryLocator: RootDirectoryLocating
+    private let signingMatcher: SigningMatching
+    private let signingCipher: SigningCiphering
+
+    public convenience init() {
+        self.init(signingFilesLocator: SigningFilesLocator(),
+                  signingMatcher: SigningMatcher(),
+                  rootDirectoryLocator: RootDirectoryLocator(),
+                  signingCipher: SigningCipher())
+    }
+
+    init(signingFilesLocator: SigningFilesLocating,
+         signingMatcher: SigningMatching,
+         rootDirectoryLocator: RootDirectoryLocating,
+         signingCipher: SigningCiphering) {
+        self.signingFilesLocator = signingFilesLocator
+        self.signingMatcher = signingMatcher
+        self.rootDirectoryLocator = rootDirectoryLocator
+        self.signingCipher = signingCipher
+    }
+
+    // MARK: - GraphMapping
+
+    public func map(graph: Graph) throws -> (Graph, [SideEffectDescriptor]) {
+        let entryPath = graph.entryPath
+        guard
+            try signingFilesLocator.locateSigningDirectory(from: entryPath) != nil,
+            let derivedDirectory = rootDirectoryLocator.locate(from: entryPath)?.appending(component: Constants.derivedFolderName)
```

Comment:
We've been creating derived directories relative to the project's directory _(not the root)_. I'd stay consistent with that and create a keychain per project in their derived directory. What do you think?

Reply:
That makes more sense, fixed 👍